packages/hemi-tunnel-actions/src/depositEth.ts

```diff
@@ -0,0 +1,113 @@
+import { EventEmitter } from 'events'
+import {
+  type Address,
+  type Chain,
+  type PublicClient,
+  type WalletClient,
+} from 'viem'
+import { writeContract } from 'viem/actions'
+
+import { l1StandardBridgeAbi } from './abis'
+import { DepositEvents } from './types'
+import {
+  getL1StandardBridgeAddress,
+  handleWaitDeposit,
+  toPromiseEvent,
+  validateInputs,
+} from './utils'
+
+const canDepositEth = async function ({
+  account,
+  amount,
+  l1Chain,
+  l1PublicClient,
+  l2Chain,
+}: {
+  account: Address
+  amount: bigint
+  l1Chain: Chain
+  l1PublicClient: PublicClient
+  l2Chain: Chain
+}): Promise<{
+  canDeposit: boolean
+  reason?: string
+}> {
+  const validation = validateInputs({
+    account,
+    amount,
+    l1Chain,
+    l2Chain,
+  })
+  if (validation) {
+    return validation
+  }
+
+  const balance = await l1PublicClient.getBalance({ address: account })
+  if (amount >= balance) {
+    return { canDeposit: false, reason: 'insufficient balance' }
+  }
+
+  return { canDeposit: true }
+}
+
+const runDepositEth = ({
+  account,
+  amount,
+  l1Chain,
+  l1PublicClient,
+  l1WalletClient,
+  l2Chain,
+}: {
+  account: Address
+  amount: bigint
+  l1Chain: Chain
+  l1PublicClient: PublicClient
+  l1WalletClient: WalletClient
+  l2Chain: Chain
+}) =>
+  async function (emitter: EventEmitter<DepositEvents>) {
+    const { canDeposit, reason } = await canDepositEth({
+      account,
+      amount,
+      l1Chain,
+      l1PublicClient,
+      l2Chain,
+    })
+
+    if (!canDeposit) {
+      // reason must be defined because canDeposit is false
+      emitter.emit('deposit-failed-validation', reason!)
+      return
+    }
+
+    emitter.emit('on-deposit')
+
+    const l1StandardBridge = getL1StandardBridgeAddress({ l1Chain, l2Chain })
+
+    // Using @ts-expect-error fails to compile so I need to use @ts-ignore
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore because it works on IDE, and when building on its own, but fails when compiling from the portal through next
+    const depositHash = await writeContract(l1WalletClient, {
+      abi: l1StandardBridgeAbi,
+      account,
+      address: l1StandardBridge,
+      // See https://github.com/ethereum-optimism/ecosystem/blob/8da00d3b9044dcb58558df28bae278b613562725/packages/sdk/src/adapters/eth-bridge.ts#L144
+      args: [200_000, '0x'],
+      chain: l1Chain,
+      functionName: 'depositETH',
+      value: amount,
+    }).catch(function (error) {
+      emitter.emit('user-signing-deposit-error', error)
+    })
+
+    await handleWaitDeposit({
+      emitter,
+      hash: depositHash,
+      publicClient: l1PublicClient,
+    })
+
+    emitter.emit('on-deposit-settled')
```

Comment:
This will be emitted even when `handleWaitDeposit` emits an error (as it just `return`s after doing so). Is that the intended behavior?

Reply:
Yes, this is an intended behavior.
This mimics the `onSettle` event from `useMutations` from `react-query` which always fires at the end, regardless of the outcome. It's used mostly for cleanup. I use it to invalidate balances so they are refetched in the background